Porting.md

```diff
@@ -0,0 +1,145 @@
+
+## Ports' Status
+The status of ports is kept in the main module [README](README.md).
+
+## Overview
+ZikiChombo sio code has a porting process that is unconventional due to the
+nature of sound support on host systems.  Usually, to port a system, on takes some
+desired cross-host functionality, such as a filesystem, and implements it with
+the OS or OS/hardware combination of interest and that's that (to make a long
+story short).
+
+For sound, the problem is that hosts usually have multiple different software
+components as entry points to playing and capturing sound, and moreover these
+entry points have some mutually exclusive properties, such as lower bounds on
+latency, availability of duplex synchronised by hardware audio clock, ability
+to be shared accross many users or applications, etc.
+
+Although the main CPU hardware architectures can certainly be a concern, it has
+not been so far in our experience.  So we focus on organizing around host,
+which is taken to be the value of runtime.GOOS in a Go program.
+
+# Hosts
+A host often has a hardware abstraction layer, hardware drivers, and some
+levels of interacting with these layers to coordinate and make secure demands
+related to playing and capturing sound on a device.
+
+To simplify the end use case, we want to make all this transparent to the
+consumer of sio, and let them simply work with sound.{Source,Sink,Duplex}.
+
+For each Host, ZikiChombo defines a default entry point to accomplish this.
+
+All entry points are available both to the consumer of ZikiChombo and either
+with ZikiChombo or as 3rd party implementations.  So if you're working with
+VoIP and have precise duplex echo cancellation needs, or a music app that
+listens and plays in real time with feedback, or writing a PulseAudio
+replacement, you're more likely to be interested in using a specialized entry
+point than the default.
+
+There is a directory ZikiChombo/sio/ports/{runtime.GOOS} for each host. 
+
+# Entry Points
+An entry point defines a proper subset of the functionality listed in the main
+[README](README.md).  ZikiChombo defines for each host (runtime.GOOS) a list of
+entry points which refer to the software layer with which the go program will
+communicate.  These are named after the respective entry points in the main
+[README](README.md).
+
+Each entry point defines a base set of functionality related to what the entry
+point supports:  device scanning, device change notifications, playback,
+capture, duplex.  Note the README excludes some functionality from some
+entry points and entry points may be incomplete.  
+
+The functionality is slightly more rich than implementing
+sio.{Capture,Play,Player,Duplex} to allow callers to achieve latency or other
+requirements.  Package sio automatically uses the Entry point to apply defaults
+so that the caller may simply call sio.{Play,Player,Capture,Duplex}.  Package
+sio also enforces that only one entry point may be in use at a time in one Go
```

Comment:
Will this prevent playback of multiple sounds simultaneously? E.g. would the end user have to mix samples in advance, or is it possible to simple:
```go
go sio.Play(snd1);
go sio.Play(snd2);
```

Reply:
Good question.  I had not intended it to prevent that.  I think some implementations may not support that, and maybe they should just enforce mutual exclusion via a lock.  Apple audio queue services has no problem with this.  But AUHAL might, it has a HOG_DEVICE setting for better performance.  ALSA hw might have problems with this but ALSA plug at least usually wouldn't.  Since it is so varied, I'm not sure what the best way is to express that in the interface.  Maybe the host.Entry implementations can just try to do the right thing (enforce exclusion, etc)?